File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int goodNodes(TreeNode* root) {
        int res = 0;
        int maximum = root->val;
        dfs(root, res, maximum);
        return res;
    }
    int dfs(TreeNode* root, int &res, int maximum){
        if(!root){
            return 0;
        }
        maximum = max(maximum, root->val);
        if( root->val >= maximum){
            res++;
        }
        dfs(root->left, res, maximum);
        dfs(root->right,res, maximum);
        return 0;
    }
};

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.res = -1001
        self.dfs(root)
        return self.res
    def dfs(self, root):
        if not root:
            return 0
        
        maxLeft = self.dfs(root.left)
        maxRight = self.dfs(root.right)
        maxLeft = max(maxLeft,0)
        maxRight = max(maxRight,0)

        self.res = max(self.res, root.val + maxLeft + maxRight)

        return root.val + max(maxLeft, maxRight)